File: buyer-ios/PoultryBuyer/Core/Error/ErrorHandler.swift
import Foundation
import Combine
import os.log

/// Centralized error handler for the application.
/// Provides error classification, logging, and user-friendly message generation.
@MainActor
final class ErrorHandler: ObservableObject {
    static let shared = ErrorHandler()

    // MARK: - Published Properties
    @Published private(set) var currentError: AppError?
    @Published private(set) var errorHistory: [ErrorLogEntry] = []

    // MARK: - Private Properties
    private let logger = Logger(subsystem: Bundle.main.bundleIdentifier ?? "PoultryBuyer", category: "ErrorHandler")
    private let maxHistorySize = 50

    // MARK: - Error Event Publisher
    private let errorSubject = PassthroughSubject<ErrorEvent, Never>()
    var errorPublisher: AnyPublisher<ErrorEvent, Never> {
        errorSubject.eraseToAnyPublisher()
    }

    private init() {}

    // MARK: - Public Methods

    /// Handles an error and converts it to an AppError
    /// - Parameters:
    ///   - error: The error to handle
    ///   - context: Optional context information
    ///   - silent: If true, won't emit error events
    /// - Returns: The converted AppError
    @discardableResult
    func handle(_ error: Error, context: ErrorContext? = nil, silent: Bool = false) -> AppError {
        let appError = AppError.from(error)

        // Log the error
        log(appError, originalError: error, context: context)

        // Emit error event
        if !silent {
            let event = ErrorEvent(error: appError, context: context, timestamp: Date())
            errorSubject.send(event)
            currentError = appError
        }

        return appError
    }

    /// Handles an API response error
    /// - Parameters:
    ///   - statusCode: HTTP status code
    ///   - message: Optional error message from server
    ///   - context: Optional context information
    /// - Returns: The converted AppError
    @discardableResult
    func handleAPIError(statusCode: Int, message: String? = nil, context: ErrorContext? = nil) -> AppError {
        let appError = AppError.from(statusCode: statusCode, message: message)

        // Log the error
        log(appError, originalError: nil, context: context)

        // Emit error event
        let event = ErrorEvent(error: appError, context: context, timestamp: Date())
        errorSubject.send(event)
        currentError = appError

        return appError
    }

    /// Clears the current error
    func clearError() {
        currentError = nil
    }

    /// Gets a user-friendly message for the error
    func getMessage(for error: AppError) -> String {
        return error.localizedMessage
    }

    /// Gets a title for the error
    func getTitle(for error: AppError) -> String {
        return error.title
    }

    /// Checks if the error is recoverable
    func isRecoverable(_ error: AppError) -> Bool {
        return error.isRecoverable
    }

    /// Checks if the error requires re-authentication
    func requiresAuthentication(_ error: AppError) -> Bool {
        return error.requiresAuthentication
    }

    // MARK: - Private Methods

    private func log(_ appError: AppError, originalError: Error?, context: ErrorContext?) {
        let entry = ErrorLogEntry(
            error: appError,
            context: context,
            timestamp: Date()
        )

        // Add to history
        errorHistory.insert(entry, at: 0)
        if errorHistory.count > maxHistorySize {
            errorHistory.removeLast()
        }

        // Log based on severity
        let message = buildLogMessage(appError: appError, originalError: originalError, context: context)

        switch appError.category {
        case .network:
            if case .noConnection = appError {
                logger.info("\(message)")
            } else {
                logger.warning("\(message)")
            }
        case .server:
            logger.error("\(message)")
        case .authentication:
            logger.warning("\(message)")
        case .client:
            logger.info("\(message)")
        case .business:
            logger.info("\(message)")
        case .unknown:
            logger.error("\(message)")
        }

        // TODO: Send to crash reporting service (Firebase Crashlytics, Sentry, etc.)
        // CrashlyticsManager.shared.logError(appError, context: context)
    }

    private func buildLogMessage(appError: AppError, originalError: Error?, context: ErrorContext?) -> String {
        var components: [String] = []

        components.append("Error: \(appError.title)")
        components.append("Code: \(appError.errorCode)")
        components.append("Message: \(appError.localizedMessage)")

        if let context = context {
            if let screen = context.screen {
                components.append("Screen: \(screen)")
            }
            if let action = context.action {
                components.append("Action: \(action)")
            }
        }

        if let originalError = originalError {
            components.append("Original: \(originalError.localizedDescription)")
        }

        return components.joined(separator: " | ")
    }
}

// MARK: - Supporting Types

/// Context information for error handling
struct ErrorContext: Sendable {
    let screen: String?
    let action: String?
    let additionalInfo: [String: String]

    init(
        screen: String? = nil,
        action: String? = nil,
        additionalInfo: [String: String] = [:]
    ) {
        self.screen = screen
        self.action = action
        self.additionalInfo = additionalInfo
    }
}

/// Event emitted when an error occurs
struct ErrorEvent: Sendable {
    let error: AppError
    let context: ErrorContext?
    let timestamp: Date
}

/// Entry in the error history log
struct ErrorLogEntry: Identifiable {
    let id = UUID()
    let error: AppError
    let context: ErrorContext?
    let timestamp: Date
}

// MARK: - Result Extension

extension Result {
    /// Converts a Result to handle errors through ErrorHandler
    @MainActor
    func handleError(
        handler: ErrorHandler = .shared,
        context: ErrorContext? = nil,
        silent: Bool = false
    ) -> Result<Success, AppError> {
        switch self {
        case .success(let value):
            return .success(value)
        case .failure(let error):
            let appError = handler.handle(error, context: context, silent: silent)
            return .failure(appError)
        }
    }
}

// MARK: - Safe API Call

/// Executes an async operation safely and returns a Result
/// - Parameters:
///   - context: Optional error context
///   - operation: The async operation to execute
/// - Returns: Result containing success value or AppError
func safeAPICall<T>(
    context: ErrorContext? = nil,
    _ operation: @escaping () async throws -> T
) async -> Result<T, AppError> {
    do {
        let result = try await operation()
        return .success(result)
    } catch {
        let appError = await ErrorHandler.shared.handle(error, context: context)
        return .failure(appError)
    }
}

/// Executes an async operation safely without returning a result
/// - Parameters:
///   - context: Optional error context
///   - onError: Callback when error occurs
///   - operation: The async operation to execute
func performSafely(
    context: ErrorContext? = nil,
    onError: ((AppError) -> Void)? = nil,
    _ operation: @escaping () async throws -> Void
) async {
    do {
        try await operation()
    } catch {
        let appError = await ErrorHandler.shared.handle(error, context: context)
        onError?(appError)
    }
}

File: buyer-android/app/src/main/java/com/poultry/buyer/core/error/ErrorHandler.kt
package com.poultry.buyer.core.error

import android.util.Log
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Centralized error handler for the application.
 * Provides error classification, logging, and user-friendly message generation.
 */
@Singleton
class ErrorHandler @Inject constructor() {

    private val _errorEvents = MutableSharedFlow<ErrorEvent>(extraBufferCapacity = 10)
    val errorEvents: SharedFlow<ErrorEvent> = _errorEvents.asSharedFlow()

    companion object {
        private const val TAG = "ErrorHandler"
    }

    /**
     * Handles an exception and converts it to an AppError.
     * Logs the error and emits an error event.
     */
    fun handleException(
        throwable: Throwable,
        context: ErrorContext? = null,
        silent: Boolean = false
    ): AppError {
        // Don't process cancellation exceptions
        if (throwable is CancellationException) throw throwable

        val appError = classifyError(throwable, context)

        // Log the error
        logError(appError, throwable)

        // Emit error event unless silent
        if (!silent) {
            _errorEvents.tryEmit(ErrorEvent(appError, context))
        }

        return appError
    }

    /**
     * Classifies a throwable into an AppError
     */
    private fun classifyError(throwable: Throwable, context: ErrorContext?): AppError {
        val networkError = NetworkError.from(throwable)

        return when (networkError) {
            is NetworkError.NoConnection -> AppError.Network(
                type = NetworkErrorType.NO_CONNECTION,
                message = networkError.message,
                isRecoverable = true
            )
            is NetworkError.Timeout -> AppError.Network(
                type = NetworkErrorType.TIMEOUT,
                message = networkError.message,
                isRecoverable = true
            )
            is NetworkError.ServerError -> AppError.Server(
                httpCode = networkError.httpCode,
                message = networkError.message,
                isRecoverable = true
            )
            is NetworkError.Unauthorized -> AppError.Auth(
                type = AuthErrorType.SESSION_EXPIRED,
                message = networkError.message
            )
            is NetworkError.Forbidden -> AppError.Auth(
                type = AuthErrorType.FORBIDDEN,
                message = networkError.message
            )
            is NetworkError.ValidationError -> AppError.Validation(
                message = networkError.message,
                fieldErrors = networkError.fieldErrors
            )
            is NetworkError.NotFound -> AppError.NotFound(
                message = networkError.message
            )
            is NetworkError.RateLimited -> AppError.RateLimited(
                retryAfterSeconds = networkError.retryAfterSeconds,
                message = networkError.message
            )
            is NetworkError.SSLError -> AppError.Network(
                type = NetworkErrorType.SSL_ERROR,
                message = networkError.message,
                isRecoverable = false
            )
            is NetworkError.Unknown -> AppError.Unknown(
                message = networkError.message,
                cause = throwable
            )
        }
    }

    /**
     * Logs the error with appropriate severity
     */
    private fun logError(error: AppError, throwable: Throwable) {
        val logMessage = buildLogMessage(error, throwable)

        when (error) {
            is AppError.Network -> {
                if (error.type == NetworkErrorType.NO_CONNECTION) {
                    Log.w(TAG, logMessage)
                } else {
                    Log.e(TAG, logMessage, throwable)
                }
            }
            is AppError.Server -> Log.e(TAG, logMessage, throwable)
            is AppError.Auth -> Log.w(TAG, logMessage)
            is AppError.Validation -> Log.d(TAG, logMessage)
            is AppError.NotFound -> Log.w(TAG, logMessage)
            is AppError.RateLimited -> Log.w(TAG, logMessage)
            is AppError.Business -> Log.i(TAG, logMessage)
            is AppError.Unknown -> Log.e(TAG, logMessage, throwable)
        }

        // TODO: Send to crash reporting service (Crashlytics, Sentry, etc.)
        // reportToCrashlytics(error, throwable)
    }

    private fun buildLogMessage(error: AppError, throwable: Throwable): String {
        return buildString {
            append("Error: ${error::class.simpleName}")
            append(" | Message: ${error.userMessage}")
            append(" | Original: ${throwable.message}")
            if (error is AppError.Server) {
                append(" | HTTP Code: ${error.httpCode}")
            }
        }
    }

    /**
     * Gets a user-friendly message for display
     */
    fun getUserMessage(error: AppError): String = error.userMessage

    /**
     * Gets a title for the error dialog
     */
    fun getErrorTitle(error: AppError): String {
        return when (error) {
            is AppError.Network -> when (error.type) {
                NetworkErrorType.NO_CONNECTION -> "No Internet Connection"
                NetworkErrorType.TIMEOUT -> "Connection Timeout"
                NetworkErrorType.SSL_ERROR -> "Security Error"
            }
            is AppError.Server -> "Server Error"
            is AppError.Auth -> when (error.type) {
                AuthErrorType.SESSION_EXPIRED -> "Session Expired"
                AuthErrorType.FORBIDDEN -> "Access Denied"
                AuthErrorType.INVALID_CREDENTIALS -> "Login Failed"
            }
            is AppError.Validation -> "Invalid Input"
            is AppError.NotFound -> "Not Found"
            is AppError.RateLimited -> "Too Many Requests"
            is AppError.Business -> "Error"
            is AppError.Unknown -> "Error"
        }
    }

    /**
     * Checks if the error is recoverable (can be retried)
     */
    fun isRecoverable(error: AppError): Boolean {
        return when (error) {
            is AppError.Network -> error.isRecoverable
            is AppError.Server -> error.isRecoverable
            is AppError.RateLimited -> true
            is AppError.Unknown -> true
            else -> false
        }
    }
}

/**
 * Sealed class representing application-level errors
 */
sealed class AppError(
    open val userMessage: String,
    open val code: Int = 0
) {
    data class Network(
        val type: NetworkErrorType,
        override val userMessage: String,
        val isRecoverable: Boolean
    ) : AppError(userMessage, type.code)

    data class Server(
        val httpCode: Int,
        override val userMessage: String,
        val isRecoverable: Boolean
    ) : AppError(userMessage, ErrorCode.SERVER_ERROR)

    data class Auth(
        val type: AuthErrorType,
        override val userMessage: String
    ) : AppError(userMessage, type.code)

    data class Validation(
        override val userMessage: String,
        val fieldErrors: Map<String, String> = emptyMap()
    ) : AppError(userMessage, ErrorCode.VALIDATION_ERROR)

    data class NotFound(
        override val userMessage: String
    ) : AppError(userMessage, ErrorCode.NOT_FOUND)

    data class RateLimited(
        val retryAfterSeconds: Int?,
        override val userMessage: String
    ) : AppError(userMessage, ErrorCode.RATE_LIMITED)

    data class Business(
        override val code: Int,
        override val userMessage: String
    ) : AppError(userMessage, code)

    data class Unknown(
        override val userMessage: String,
        val cause: Throwable? = null
    ) : AppError(userMessage, ErrorCode.UNKNOWN)
}

enum class NetworkErrorType(val code: Int) {
    NO_CONNECTION(ErrorCode.NO_CONNECTION),
    TIMEOUT(ErrorCode.TIMEOUT),
    SSL_ERROR(ErrorCode.SSL_ERROR)
}

enum class AuthErrorType(val code: Int) {
    SESSION_EXPIRED(ErrorCode.UNAUTHORIZED),
    FORBIDDEN(ErrorCode.FORBIDDEN),
    INVALID_CREDENTIALS(ErrorCode.UNAUTHORIZED)
}

/**
 * Context information for error handling
 */
data class ErrorContext(
    val screen: String? = null,
    val action: String? = null,
    val additionalInfo: Map<String, Any> = emptyMap()
)

/**
 * Event emitted when an error occurs
 */
data class ErrorEvent(
    val error: AppError,
    val context: ErrorContext?
)

/**
 * Extension function to execute a block safely and return a Result
 */
suspend fun <T> safeApiCall(
    errorHandler: ErrorHandler,
    context: ErrorContext? = null,
    block: suspend () -> T
): Result<T> {
    return try {
        Result.success(block())
    } catch (e: Exception) {
        val appError = errorHandler.handleException(e, context)
        Result.failure(AppException(appError))
    }
}

/**
 * Exception wrapper for AppError
 */
class AppException(val appError: AppError) : Exception(appError.userMessage)

/**
 * Extension function to get AppError from Result failure
 */
fun <T> Result<T>.getAppError(): AppError? {
    return exceptionOrNull()?.let {
        if (it is AppException) it.appError else null
    }
}
